components/client/src/test/unit/java/com/hotels/styx/client/TransportTest.java

```diff
@@ -67,111 +70,150 @@ public void setUp() {
     }
 
     @Test
-    public void releasesIfRequestIsUnsubscribedAfterHeaders() {
+    public void returnsConnectionBackToPool_delayedResponseError() {
         Connection connection = mockConnection(responseProvider);
         ConnectionPool pool = mockPool(connection);
 
-        HttpTransaction transaction = transport.send(request, Optional.of(pool), APP_ID);
-
-        transaction.response().subscribe(subscriber);
+        transport.send(request, Optional.of(pool), APP_ID)
+                .response()
+                .subscribe(subscriber);
 
         responseProvider.onNext(response);
+        responseProvider.onError(new RuntimeException("oh dear ..."));
 
-        subscriber.unsubscribe();
-        verify(pool).closeConnection(any(Connection.class));
+        subscriber.getOnNextEvents().get(0).consume();
+
+        verify(pool).returnConnection(any(Connection.class));
     }
 
     @Test
-    public void closesIfObservableUnsubscribedBeforeHeaders() {
-        ConnectionPool pool = mockPool(mockConnection(PublishSubject.<HttpResponse>create()));
+    public void terminatesConnection_emptyHeadersObservable() {
+        Connection connection = mockConnection(Observable.empty());
+        ConnectionPool pool = mockPool(connection);
 
-        Subscription subscription = transport.send(request, Optional.of(pool), APP_ID)
+        transport.send(request, Optional.of(pool), APP_ID)
                 .response()
-                .subscribe(new TestSubscriber<>());
+                .subscribe(subscriber);
 
-        verify(pool, never()).closeConnection(any(Connection.class));
+        assertThat(subscriber.getOnNextEvents().isEmpty(), is(true));
+        assertThat(subscriber.getOnCompletedEvents().size(), is(1));
 
-        subscription.unsubscribe();
         verify(pool).closeConnection(any(Connection.class));
     }
 
     @Test
-    public void closesIfObservableUnsubscribedAfterHeaders() {
-        ConnectionPool pool = mockPool(mockConnection(responseProvider));
+    public void releasesConnectionWhenRequestFailsBeforeHeaders() {
+        Connection connection = mockConnection(Observable.error(new RuntimeException()));
+        ConnectionPool pool = mockPool(connection);
 
-        Subscription subscription = transport.send(request, Optional.of(pool), APP_ID)
+        transport.send(request, Optional.of(pool), APP_ID)
                 .response()
-                .subscribe(new TestSubscriber<>());
+                .subscribe(subscriber);
 
-        responseProvider.onNext(response);
-        verify(pool, never()).closeConnection(any(Connection.class));
+        assertThat(subscriber.getOnNextEvents().isEmpty(), is(true));
+        assertThat(subscriber.getOnErrorEvents().size(), is(1));
 
-        subscription.unsubscribe();
         verify(pool).closeConnection(any(Connection.class));
     }
 
+    @Test
+    public void terminatesConnection_bodyIsUnsubscribed() {
```

Comment:
Again.

Reply:
Will rename to `returnsConnectionBackToPoolDueToDelayedResponseError`.